food_delivery/foodapp/views.py

```diff
@@ -1,3 +1,453 @@
+        serializer = OrderSerializer(order)
+        return Response(serializer.data)
+
+    def put(self, request, pk):
+        order = self.get_object(pk)
+        if order is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+        serializer = OrderSerializer(order, data=request.data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data)
+        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+
+    def delete(self, request, pk):
+        order = self.get_object(pk)
+        if order is None:
+            return Response(status=status.HTTP_404_NOT_FOUND)
+        order.delete()
+        return Response(status=status.HTTP_204_NO_CONTENT)
+
+
+class OrderActionsView(APIView):
+    """
+    Handles custom actions for specific orders: calculating delivery time, marking as delivered, canceling.
+    """
+
+    def get_object(self, pk):
+        try:
+            return Order.objects.get(pk=pk)
+        except Order.DoesNotExist:
+            return None
+
+    def calculate_delivery_time(self, order):
+        """
+        A function to calculate the estimated delivery time for an order.
+        It uses the geolocation of the restaurant and the customer.
+        """
+        geolocator = Nominatim(user_agent="food_delivery_app")
+        restaurant_location = geolocator.geocode(order.restaurant.address)
+        customer_location = geolocator.geocode(order.customer.current_location)
+
+        if restaurant_location and customer_location:
+            restaurant_coords = (restaurant_location.latitude, restaurant_location.longitude)
+            customer_coords = (customer_location.latitude, customer_location.longitude)
+            distance = geodesic(restaurant_coords, customer_coords).km
+            
+            # Assuming average delivery speed is 40 km/h
+            estimated_time = distance / 40 * 60  # Time in minutes
+            return datetime.now() + timedelta(minutes=estimated_time)
+        return None
+
+    def get(self, request, pk, action):
+        """
+        GET method for actions like 'estimated_time'.
+        Accessible at: /orders/{pk}/actions/{action}/ (e.g., /orders/1/actions/estimated_time/)
+        """
+        order = self.get_object(pk)
+        if order is None:
+            return Response({"error": "Order not found"}, status=status.HTTP_404_NOT_FOUND)
+
+        if action == 'estimated_time':
+            estimated_time = self.calculate_delivery_time(order)
+            if estimated_time:
+                return Response({"estimated_delivery_time": estimated_time.strftime("%Y-%m-%d %H:%M:%S")}, status=status.HTTP_200_OK)
+            return Response({"error": "Could not calculate delivery time"}, status=status.HTTP_400_BAD_REQUEST)
+
+        return Response({"error": "Invalid action"}, status=status.HTTP_400_BAD_REQUEST)
+
+    def post(self, request, pk, action):
```

Comment:
Django rest framework has action decorators. it's little advanced topic but will be helpful. read about it.

Reply:
ok sir